assembly/src/sema/passes/const_eval.rs

```diff
@@ -0,0 +1,77 @@
+use crate::{
+    ast::*,
+    sema::{AnalysisContext, SemanticAnalysisError},
+    Felt, Span, Spanned,
+};
+use core::ops::ControlFlow;
+
+/// This visitor evaluates all constant expressions and folds them to literals.
+pub struct ConstEvalVisitor<'analyzer> {
+    analyzer: &'analyzer mut AnalysisContext,
+}
+
+impl<'analyzer> ConstEvalVisitor<'analyzer> {
+    pub fn new(analyzer: &'analyzer mut AnalysisContext) -> Self {
+        Self { analyzer }
+    }
+}
+
+impl<'analyzer> ConstEvalVisitor<'analyzer> {
+    fn eval_const<T>(&mut self, imm: &mut Immediate<T>) -> ControlFlow<()>
+    where
+        T: TryFrom<u64>,
+    {
+        match imm {
+            Immediate::Value(_) => ControlFlow::Continue(()),
+            Immediate::Constant(ref name) => {
+                let span = name.span();
+                match self.analyzer.get_constant(name) {
+                    Ok(value) => match T::try_from(value.as_int()) {
+                        Ok(value) => {
+                            *imm = Immediate::Value(Span::new(span, value));
```

Comment:
I just noticed this is inlining all constants into the code. And it effectively erases all constants from the parsed module:

https://github.com/0xPolygonMiden/miden-vm/blob/cb162987947c3a58267ad289be141820ce8c7900/assembly/src/tests.rs#L401-L404

Do we have to do that? I think it would be really beneficial to have the constants still available, at least the final result. This would allow us to remove the need to redefine constants in MASM and Rust:

```
# The memory address at which the chain data section begins
const.CHAIN_MMR_PTR=300
```

[ref](https://github.com/0xPolygonMiden/miden-base/blob/57185284e8db55d4f7f74781072309e131adad45/miden-lib/asm/miden/kernels/tx/memory.masm#L81-L82)

```rust
/// The memory address at which the chain data section begins
pub const CHAIN_MMR_PTR: MemoryAddress = 300;
```

[ref](https://github.com/0xPolygonMiden/miden-base/blob/57185284e8db55d4f7f74781072309e131adad45/miden-lib/src/transaction/memory.rs#L108-L109)

Reply:
In theory const evaluation can be delayed, and/or provided with some kind of implicit prelude of predefined constants as part of the parser/assembler context. However, I would caution against this: referencing constants in the MASM source code that have no definitions in the same file, or at least an imported module, would be very confusing to the reader - especially if the source of those constant definitions is dynamic in nature (i.e. provided on the fly to the parser/assembler). It would also preclude the ability to use IDE tools to navigate to the definition of those items. We don't have a language server for MASM yet, but I imagine we'd want one eventually - but things like this make such tools less useful.

That said, I think we can achieve the desired results a slightly different way, using a combination of two things:

1.) Support importing constants from other modules, i.e. `use.foo` allows referencing constants defined in `foo` via `foo::BAR` syntax, which could be supported in const expressions as well, e.g. `const.A = foo::BAR + 1`. With this, all common constants could be moved to a single module/namespace, and imported where needed.

2.) Allow gathering constant definitions from a parsed module, even after they have been evaluated to their concrete values, so that they can be introspected later. In other words, you parse the `foo` module, and while the actual AST is all literals, you can query the parsed `Module` for the concrete definition of `BAR`, something like `module.get_constant("BAR")`, which would return a `Constant` representing that definition. 

This would allow you to do some useful things. For example, write a build script that parses a MASM module containing constants you want to use in Rust, and generate a Rust module in `OUT_DIR` containing those constants. Rather than needing to define the constants twice, once in Rust and once in MASM, you define them once in MASM, and then generate the Rust constants. This approach also works well both with Rust's language server, and with a theoretical future MASM language server.

What do you think? Anything missing there that you think wouldn't be handled/enabled by those two features?

Reply:
Yes, I was trying to suggest option 2. 👍

Reply:
follow up: if we do implement the ability to introspect the constants from the parsed module. IMO we should forbid the usage of immediates for instructions like `emit`, `trace`, `assert.err`. My rationale is that we need a way of inspecting all of these values from the Host, and forcing the usage of constants will consolidate everything under a single method, instead of having to come up with something else for `emit.2` for example (or even worse, not come up with anything, and have to sync things by hand).

Reply:
I wouldn't _forbid_ literals for those instructions, but emitting a warning does seem useful to me (which could become an error with a compilation option like `warnings_as_errors`).

So internally we could enforce using named constants for those things, while not forcing one to do so during development/prototyping/debugging/whatever.

Reply:
Fair enough, as long as we don't have to manually keep these things in sync for our stuff I'm happy.

Are we planning on creating some CLI tools to report issues and integrate with CI?

Reply:
We could use either the `miden` CLI or the compiler for that. The compiler supports `cargo check`-like builds, but we could support that in `miden` easily enough (i.e. compile, but don't actually emit any artifacts).

On the topic of diagnostics: I'd also like - in a later PR(s) - to make some further refinements to a few aspects of how source code is managed/represented/parsed, and how we structure diagnostics that would enable better integration with the compiler, allow introduction of a MASM language server, provide a better way to support source-level debug information (e.g. source maps). What I implemented in this PR provides a simple implementation of the more important functionality (emitting diagnostics with source snippets, unifying how identifiers are represented, adding some barebones interning support), but there is much room for improvement. The key thing was to get something in place that would make later improvements easier to implement, and less drastic.

The ultimate goal from my perspective is to be able to build really good debugging/development tooling for working with Miden Assembly, MAST, and programs executed by the VM, but it's definitely an iterative process.